### PR TITLE
Recognize date/time formats

### PR DESCRIPTION
**Pull request by retokohli**
Recognize and handle known date and/or time formats when data is sent to the API.
Correctly deals with empty and DateTime values as well.

### CODE REVIEW

File: core/DataSource/Model/Entity/DoctrineRepository.class.php [hunk 1]
```diff
@@ -406,8 +406,26 @@ protected function setEntityData($entity, $data) {
...
+            if ($fieldDefinition['type'] === 'datetime'
+                || $fieldDefinition['type'] === 'timestamp'
+                || $fieldDefinition['type'] === 'date'
+                || $fieldDefinition['type'] === 'time'
+            ) {
+                if (!$data[$name]) {
+                    // Empty values must be NULL, or Doctrine will fail
+                    $data[$name] = null;
+                }
+                if ($data[$name] && !is_a($data[$name], 'DateTime')) {
```
Comment (MichaelRitter) [inline core/DataSource/Model/Entity/DoctrineRepository.class.php]: Please use instanceof as `is_a()` might get deprecated (again): https://externals.io/message/103779#103805

File: core/DataSource/Model/Entity/DoctrineRepository.class.php [hunk 2]
```diff
@@ -406,8 +406,26 @@ protected function setEntityData($entity, $data) {
             }
 
             $fieldDefinition = $entityClassMetadata->getFieldMapping($name);
-            if ($fieldDefinition['type'] == 'datetime') {
-                $data[$name] = new \DateTime($data[$name]);
+            if ($fieldDefinition['type'] === 'datetime'
+                || $fieldDefinition['type'] === 'timestamp'
+                || $fieldDefinition['type'] === 'date'
+                || $fieldDefinition['type'] === 'time'
```
Comment (MichaelRitter) [inline core/DataSource/Model/Entity/DoctrineRepository.class.php]: Please use a `switch`-statement

File: core/DataSource/Model/Entity/DoctrineRepository.class.php [hunk 3]
```diff
@@ -406,8 +406,26 @@ protected function setEntityData($entity, $data) {
...
+            ) {
+                if (!$data[$name]) {
+                    // Empty values must be NULL, or Doctrine will fail
+                    $data[$name] = null;
+                }
+                if ($data[$name] && !is_a($data[$name], 'DateTime')) {
+                    // Throws on invalid input
+                    $datetime = new \DateTime($data[$name]);
+                    // Not necessary, as the database ignores unused parts anyway.
+                    //if ($fieldDefinition['type'] === 'date') {
```
Comment (MichaelRitter) [inline core/DataSource/Model/Entity/DoctrineRepository.class.php]: Please remove the commented code. Update or remove the comment.

File: core/DataSource/Model/Entity/DoctrineRepository.class.php (hunk 2 above)
Comment (retokohli) [reply core/DataSource/Model/Entity/DoctrineRepository.class.php]: Done.  Included the case 'array' as well.

File: core/DataSource/Model/Entity/DoctrineRepository.class.php (hunk 1 above)
Comment (retokohli) [reply core/DataSource/Model/Entity/DoctrineRepository.class.php]: Done.

File: core/DataSource/Model/Entity/DoctrineRepository.class.php (hunk 3 above)
Comment (retokohli) [reply core/DataSource/Model/Entity/DoctrineRepository.class.php]: Done.  Also removed the unnecessary $datetime variable.

**Comment by retokohli** [comment]
Please see latest commits.

**Pull request merged**